### PR TITLE
Bug 1702832: pkg/operator: poll during master MCP sync

### PR DESCRIPTION
**Pull request by runcom**
The master MCP we sync in the operator can flip Degraded=True too early
if it's still rolling out. This patch introduces back a 10 minutes
polling to give it some time to correctly roll. We also now bubble up
any error to the CO's Extension field along with a full MCPs status. The
net result of this patch is that we only flip Degraded=True if: 1)
10m are up and we do have a real error preventing roll out 2) 10m are up
and we have transient errors slowing down the roll out. In both case we
report what's happening at every iteration of the sync during the 10
mins poll w/o flipping Degraded=True (as most errors are transient
anyway). If we get a permanent error during sync, we still wait 10 mins
before reporting Degraded but the error can be seen immediately in the
clusteroperator status extension.

### CODE REVIEW

**Comment by runcom** [comment]
let's see what e2e-aws-upgrade has to say about this also

**Comment by runcom** [comment]
There are some gotchas with this PR tho, most notably, 10minutes timeout has been chosen by what was there previously but if today we have a 100 nodes master pool, 10m aren't obviously enough (and we never really tested a performance run on such a huge pool I guess).

**Comment by cgwalters** [comment]
> today we have a 100 nodes master pool

We only support 3 masters today, I don't think the etcd upstream suggests anything higher than 7 https://coreos.com/etcd/docs/latest/v2/admin_guide.html

**Comment by cgwalters** [comment]
OK the commit message keeps talking about "master" as if it's special cased but...we didn't special case it in any way before this PR, and we still aren't with it right?

**Comment by cgwalters** [comment]
(And we know we need to apply OS updates to the master pool first and we're not doing that yet)

**Comment by runcom** [comment]
> OK the commit message keeps talking about "master" as if it's special cased but...we didn't special case it in any way before this PR, and we still aren't with it right?

the syncRequiredConfigPool only takes care of masters (since the master pool has a special annotation) so we only care about master effectively

**Comment by cgwalters** [comment]
So...now we end up blocking in `sync()` for the pools; I had thought the operator was responsible for e.g. regenerating the `controllerconfig` but...oh yes it is, it templates it from the renderconfig looks like.  So if...hm, I guess we don't care about immediately reflecting anything in the render config.

**Comment by runcom** [comment]
> So...now we end up blocking in `sync()` for the pools

this is what we were doing in the past as well but we failed at providing why we were still syncing when we got some transient/permanent error. The change today is introducing that back but bubbling up any error which is what we wanted at first removing the polling.

**Comment by cgwalters** [comment]
> the syncRequiredConfigPool only takes care of masters (since the master pool has a special annotation) so we only care about master effectively

Ahh...yes I see that now.  Wait but hm that makes me wonder about how we're rolling up any failures in the worker pool up to the operator.  Anyways, that doesn't need to block this PR.  LGTM from me, will let Kirsten do any other review and take final merge.

**Comment by runcom** [comment]
> Ahh...yes I see that now. Wait but hm that makes me wonder about how we're rolling up any failures in the worker pool up to the operator.

so, we bubble the error from workers in the ClusterOperator.Status.Extension field but we don't set Degraded but yeah, at some point maybe we'll need something like that for worke

**Comment by runcom** [comment]
repushing to trigger another e2e-aws-upgrade job

File: pkg/operator/sync.go [hunk 1]
```diff
@@ -330,17 +327,51 @@ func (optr *Operator) syncRequiredMachineConfigPools(config renderConfig) error
 		}
 		return false
 	}
-	for _, pool := range pools {
-		if err := isMachineConfigPoolConfigurationValid(pool, version.Version.String(), optr.mcLister.Get); err != nil {
-			return fmt.Errorf("pool %s has not progressed to latest configuration: %v, retrying", pool.Name, err)
+
+	var lastErr error
+	if err := wait.Poll(time.Second, 10*time.Minute, func() (bool, error) {
```
Comment (kikisdeliveryservice) [inline pkg/operator/sync.go]: is 10 too much/too little?

Comment (runcom) [reply pkg/operator/sync.go]: looks like it's fine according to Derek so I'd say it's good enough (and what was there before I removed it some months ago)

Comment (wking) [reply pkg/operator/sync.go]: Cross-ref https://github.com/openshift/machine-config-operator/pull/554/files#diff-fa45321336db7ad1cedc28bf643a4f97L326

Comment (kikisdeliveryservice) [reply pkg/operator/sync.go]: gotcha and thanks @wking !

**Comment by kikisdeliveryservice** [comment]
is this ready or do you need to test more?

**Comment by runcom** [comment]
> is this ready or do you need to test more?

waiting to hear from upgrade job

**Comment by runcom** [comment]
second upgrade job in a row passing, let me repush again to trigger another and then we can merge

**Comment by kikisdeliveryservice** [comment]
sounds good!

**Comment by runcom** [comment]
green

**Pull request merged**